### PR TITLE
Ui improvements stage ii

### PR DESCRIPTION
**Pull request by jantocv**
## Description
Final stage of UI library refactoring & improving.

## Changes applied:
 - Unit test coverage. 
 image

 - New family members:
   - New `CopyToClipboard` button.
   - Added `Toggle Hook` & `copy-to-clipboard` utilities.
      - Removed `withCopyBehaviour` HOC in favour of `copy-to-clipboard` utils and `CopyToClipboard` button.

 - Component refactors from class to functional 💪 
 - Natural renaming of some components:
   -  `ListCopyable` to `CopyableList`.
   - `ButtonWrapped` to `WrappedButton`.
   - ... etc

- Several refactors in almost all components ⛑️ 
   - Refactors for Field, TextField, PhoneField, DateField & BankAccountField:
      - Removing excesive use of `refs` on them.
      - Added `useMemo` hook.

- Changed `index.jsx` entry point of each component folder to component file name (`MyComponent.jsx`) 🔖 

- Updated READMEs of each component 🗒️ 

- 🥳 **New** Search Input in UI website (cool thing since we are not using storybook... at least for now). Documentation is up-to-date and there are some basic examples for each component. Search can be done by group or by specific component...

filters

----

## Checks
- [x] Requires lib version update

## Versioning impact
- [x] **MAJOR** incompatible API changes.

_Note: Please do not try to refactor this at home_

### CODE REVIEW

**Comment by rahulpsingh** [comment]
One initial thing which can be improved is, when we have file structure for example `src/Button/Button.jsx` and `Button/Button.test.js`, including a `package.json` with:

```
{
  "name": "Button",
  "version": "0.0.0",
  "private": true,
  "main": "./Button.jsx"
}
```

Can fix this from importing the Button component as `./Button/Button` from outside the component, it kind of looks anti pattern. What do you think?

I'll be taking a look at other things now!!

File: src/forms/PhoneField/PhoneField.test.jsx [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    it('Should not show error label if given value is valid', () => {
+        const {getByTestId} = render(<PhoneField  {...defaultProps} value='5512341234'/>);
+        const noErrorLabel = !getByTestId('phone-field-label').className.includes('error');
+        expect(noErrorLabel).toBeTruthy();
+    });
+
+    it('Should show error label if given value is NOT valid', () => {
+        const {getByTestId} = render(<PhoneField  {...defaultProps} errorLabel={'Error!'} value='551234123234324'/>);
+        const errorLabel = getByTestId('phone-field-label').className.includes('error');
+        expect(errorLabel).toBeTruthy();
```
Comment (JoseOrtiz) [inline src/forms/PhoneField/PhoneField.test.jsx:26]: Could you check if it is `Error!` what's displayed, as well?

File: src/index.js [hunk 2]
```diff
@@ -1,28 +1,28 @@
...
-export { default as Scroll } from './structure/Scroll'
-export { default as Wrapper } from './structure/Wrapper'
-export { default as Heading } from './typography/Heading'
-export { default as Text } from './typography/Text'
-export { default as TextField } from './forms/TextField'
-export { default as TextFieldRule } from './forms/TextFieldRule'
-export { default as TopBar } from './navigation/TopBar'
-export { default as DateField } from './forms/DateField'
-export { default as CopyableList } from './lists/CopyableList'
+export { default as Button } from './buttons/Button/Button';
```
Comment (JoseOrtiz) [inline src/index.js:1]: Why did you change the order? Is this supposed to have a special sort?

File: src/lists/ContentItem/ContentItem.test.jsx [hunk 3]
```diff
@@ -10,66 +10,66 @@ describe('UI/lists/ContentItem', () => {
 		meta: 'META',
 		textColor: 'TEXTCOLOR',
 		onClick: () => {
-			console.log('test')
-			return 'bu'
+			console.log('test');
```
Comment (JoseOrtiz) [inline src/lists/ContentItem/ContentItem.test.jsx:13]: Do we still need this `console.log`?

File: src/lists/ListItem/ListItem.jsx [hunk 4]
```diff
@@ -1,44 +1,36 @@
...
-	checked,
-	dataTestId,
-}) => (
+const ListItem = ({ title, icon, iconColor, description, className, onLink, linkText, onChange, checked, dataTestId }) => (
 	<li className={`${CSS.item} ${className}`} data-testid={dataTestId}>
-		{ (icon) ? <div data-testid={`${dataTestId}-icon`} className={`${CSS.media} ${ (iconColor ? CSS[iconColor] : '')}`}> {icon} </div> : null }
+		{icon ? (
+			<div data-testid={`${dataTestId}-icon`} className={`${CSS.media} ${iconColor ? CSS[iconColor] : ''}`}>
+				{' '}
+				{icon}{' '}
```
Comment (JoseOrtiz) [inline src/lists/ListItem/ListItem.jsx:13]: Consolidate this 2 lines into one
```suggestion
				{` ${icon} `}
```

File: src/lists/ListItem/ListItem.jsx [hunk 5]
```diff
@@ -1,44 +1,36 @@
...
 			<Heading type="txt-medium">{title}</Heading>
-			{ (description) ? <Text dataTestId={`${dataTestId}-description`}>{description}</Text> : null}
-			{ (onLink) ? (
+			{description ? <Text dataTestId={`${dataTestId}-description`}>{description}</Text> : null}
+			{onLink ? (
 				<span
 					data-testid={`${dataTestId}-heading-link`}
 					onClick={() => {
-						onLink()
+						onLink();
```
Comment (JoseOrtiz) [inline src/lists/ListItem/ListItem.jsx:23]: This could be used as a direct prop
```suggestion
					onClick={onLink}
```

**Comment by jasonadt** [comment]
> One initial thing which can be improved is, when we have file structure for example `src/Button/Button.jsx` and `Button/Button.test.js`, including a `package.json` with:
> 
> ```
> {
>   "name": "Button",
>   "version": "0.0.0",
>   "private": true,
>   "main": "./Button.jsx"
> }
> ```
> 
> Can fix this from importing the Button component as `./Button/Button` from outside the component, it kind of looks anti pattern. @jantocv What do you think?
> 
> I'll be taking a look at other things now!!

Regarding `package.json` as we had previously discussed: let's please have only one package.json at the root of the project. I would prefer that we use explicit imports as much as possible. Let me know if you'd like to discuss further.

File: docs/Navigation/Navigation.jsx [hunk 6]
```diff
@@ -1,54 +1,81 @@
...
+import { useState } from 'preact/hooks';
 import { Link } from 'preact-router';
-import 'src/Theme'
-import CSS from './Navigation.css'
 
-import Scroll from 'src/structure/Scroll'
-import Logo from  'docs/images/img_logo'
+import 'src/Theme/Theme';
+import CSS from './Navigation.css';
+import Scroll from 'src/structure/Scroll/Scroll';
```
[the remark this answers is not in the document]
Comment (jasonadt) [reply docs/Navigation/Navigation.jsx:6]: Let's please prefer to use explicit imports as much as possible.

**Comment by jasonadt** [comment]
@jantocv when are you planning on applying our new ESLint configuration to this repo? Perhaps it would make sense to do that as part of this PR? I'm fine with a separate one too if you'd prefer.

**Comment by jantocv** [comment]
> One initial thing which can be improved is, when we have file structure for example `src/Button/Button.jsx` and `Button/Button.test.js`, including a `package.json` with:
> 
> ```
> {
>   "name": "Button",
>   "version": "0.0.0",
>   "private": true,
>   "main": "./Button.jsx"
> }
> ```
> 
> Can fix this from importing the Button component as `./Button/Button` from outside the component, it kind of looks anti pattern. @jantocv What do you think?
> 
> I'll be taking a look at other things now!!

it can be discussed, sure, but I am not understanding the advantage. A part from the anti pattern, I mean, what is the difference between doing it with separate package.json and doing it like we are doing it now?

File: docs/App.jsx [hunk 7]
```diff
@@ -152,176 +174,27 @@ export default class App extends Component {
 						<LiveError />
 						<LivePreview />
 					</LiveProvider>
-				)
+				);
 			} else if (props.metaString == 'readonly') {
-				return <div>{props.children}</div>
+				return <div>{props.children}</div>;
```
[the remark this answers is not in the document]
Comment (jantocv) [reply docs/App.jsx:179]: Do not know being honest. I did not changed too much the structure of them, I am going to check ;)

File: src/forms/BankAccountField/BankAccountField.jsx [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+	};
+
+	const accountNumberFormatter = ({ accountNumber, template }) => {
+		let accountDigitIndex = 0;
+		const generateAccountValue = (prevResult, patternCharacter) => {
+            if (!accountNumber[accountDigitIndex]) { return prevResult; }
+            let nextCharacter = (patternCharacter === digitPattern) ? accountNumber[accountDigitIndex] : patternCharacter;
+            if (patternCharacter === digitPattern) { accountDigitIndex++; }
+            const newResult = `${prevResult}${nextCharacter}`;
+            return newResult;
```
[the remark this answers is not in the document]
Comment (jantocv) [reply src/forms/BankAccountField/BankAccountField.jsx:50]: Could be, thank you! My editor its starting doing weird things... :(

File: src/forms/PhoneField/PhoneField.test.jsx (hunk 1 above)
Comment (jantocv) [reply src/forms/PhoneField/PhoneField.test.jsx:26]: Nice one! sure :) thanks!

File: src/index.js (hunk 2 above)
Comment (jantocv) [reply src/index.js:1]: Nope, it was only to group them a bit by "folder" (forms, lists, etc...)

File: src/lists/ContentItem/ContentItem.test.jsx (hunk 3 above)
Comment (jantocv) [reply src/lists/ContentItem/ContentItem.test.jsx:13]: Of course not! thanks!

**Comment by jantocv** [comment]
> @jantocv when are you planning on applying our new ESLint configuration to this repo? Perhaps it would make sense to do that as part of this PR? I'm fine with a separate one too if you'd prefer.

I do not have a specific date for that -> ASAP.  Of course it can be taken in this "series of PRs", lets do a STAGE III - PR  :)

File: docs/App.jsx [hunk 9]
```diff
@@ -142,8 +163,9 @@ export default class App extends Component {
 				TextFieldRule,
 				DateField,
 				CopyableList,
+				Toggle,
 				...Icons,
-			}
+			};
 
 			if (props.metaString == null) {
```
[the remark this answers is not in the document]
Comment (jantocv) [reply docs/App.jsx:170]: I did not take too much care on the "UI web" because it is not relevant to other repos... it is just like a web to show the components we have... maybe it will be removed in favour of Storybook some day 🤷‍♂️ I think it does not make sense to spend too much time refactoring it. What do u think?

**Comment by rahulpsingh** [comment]
> it can be discussed, sure, but I am not understanding the advantage. A part from the anti pattern, I mean, what is the difference between doing it with separate package.json and doing it like we are doing it now?

Not much of difference @jantocv . But as we will be using single package.json per repo, let's try to find a way to export/import/organise it in better way. Maybe, `index.jsx` and `index.test.jsx`? Let's discuss this once!

**Comment by jasonadt** [comment]
> > it can be discussed, sure, but I am not understanding the advantage. A part from the anti pattern, I mean, what is the difference between doing it with separate package.json and doing it like we are doing it now?
> 
> Not much of difference @jantocv . But as we will be using single package.json per repo, let's try to find a way to export/import/organise it in better way. Maybe, `index.jsx` and `index.test.js`? Let's discuss this once!

As discussed over Slack: we will prefer explicit imports, which means that we have only **one** `package.json` and only **one** `index.js` at the root of each project.

File: src/forms/Toggle/Readme.md [hunk 10]
```diff
@@ -0,0 +1,22 @@
+# Toggle
+
+A simple toggle input: The control of this component should be handled in parent.
```
[the remark this answers is not in the document]
Comment (jantocv) [reply src/forms/Toggle/Readme.md:3]: 🇺🇸 🇬🇧 Thanks!

File: src/buttons/Button/Button.jsx [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+	const testId = dataTestId || defaultTestId;
+
+	if (isButton) {
+		return (
+			<button 
+				data-testid={`${testId}-button`} 
+				type="button" id={id} 
+				onTouchStart={() => setIsPressed(!isPressed)} 
+				onTouchEnd={() => setIsPressed(!isPressed)} 
+				onClick={e => onHandleClick(e)}
```
Comment (jasonadt) [inline src/buttons/Button/Button.jsx:36]: (nit) please simplify this line to: 

```js
onClick={onHandleClick} 
```

File: src/buttons/Button/Button.jsx [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+	const isButton = !href;
+	const defaultTestId = isButton ? 'button' : 'button-link';
+	const testId = dataTestId || defaultTestId;
+
+	if (isButton) {
+		return (
+			<button 
+				data-testid={`${testId}-button`} 
+				type="button" id={id} 
+				onTouchStart={() => setIsPressed(!isPressed)}
```
Comment (jasonadt) [inline src/buttons/Button/Button.jsx:34]: (DRY) please refactor this arrow-function into a new `const` above line: 22 and then reuse it here and everywhere else you have `() => setIsPressed(!isPressed)`

File: src/buttons/CopyToClipboard/CopyToClipboard.jsx [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+	const handleOnClick = textToCopy => {
+		setCopyStatus(copyStatusLookup.copying);
+		setTimeout(() => {
+			copyToClipboard(textToCopy, afterCopy);
+		}, copyTimeout);
+	};
+
+	return (
+		<button data-testid={`${dataTestId}-button`} className={`${CSS.clipboard} ${CSS[copyStatus.class]}`} onClick={() => handleOnClick(toCopyValue)}>
```
Comment (jasonadt) [inline src/buttons/CopyToClipboard/CopyToClipboard.jsx:43]: (nit) the `onClick` arrow-function is missing curly braces around its body (and a semicolon), please add them.

File: src/forms/PhoneField/PhoneField.jsx [hunk 14]
```diff
@@ -0,0 +1,95 @@
+import {useState, useEffect, useRef, useMemo} from 'preact/hooks';
+import PropTypes from 'prop-types'
+import Metadata from '../../../data/libphonenumber-metadata.min.json'
```
Comment (jasonadt) [inline src/forms/PhoneField/PhoneField.jsx:3]: (nit) this line should be grouped with the other project-level imports, below.

File: src/forms/SelectionList/SelectionList.jsx [hunk 15]
```diff
@@ -0,0 +1,109 @@
+import { useState, useEffect, useMemo } from 'preact/hooks';
+import PropTypes from 'prop-types';
+
+import Heading from '../../typography/Heading/Heading';
+import CSS from './styles.css';
+
+const SelectionList = ({ className, dataTestId, value, title, options, id, onChange }) => {
+	const [currentValue, setCurrentValue] = useState(null);
+	const delayCallback = 250;
```
Comment (jasonadt) [inline src/forms/SelectionList/SelectionList.jsx:9]: (naming) please rename this to: `callbackDelayInMilliseconds` and move it to a new line above line: 7.

File: src/forms/SelectionList/SelectionList.jsx [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+			setCurrentValue(value);
+			return;
+		}
+		const defaultCheckedOptionIndex = options.findIndex(({ selected }) => selected);
+		setCurrentValue(defaultCheckedOptionIndex !== -1 ? options[defaultCheckedOptionIndex].value : null);
+	}, [value, options]);
+
+	const handleOnClick = ({ selected, label, value: optionValue, children, onClick }, index) => {
+		if (onClick) {
+			setTimeout(() => onClick(optionValue, index), delayCallback);
```
Comment (jasonadt) [inline src/forms/SelectionList/SelectionList.jsx:22]: (nit) please enclose this arrow-function body within curly-braces.

Same comment for line: 25, below.

File: src/forms/SelectionList/SelectionList.jsx [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+						return (
+							<li 
+								data-testid={`${dataTestId}-option`} 
+								key={index} 
+								data-checked={checked} 
+								className={`${CSS.item} ${checked ? CSS.checked : ''}`} 
+								onClick={event => handleOnClick({ selected, label, value: optionValue, children, onClick }, index)}
+							>
+								<input 
+									type="radio" c
```
Comment (jasonadt) [inline src/forms/SelectionList/SelectionList.jsx:52]: Please move the `c` at the end of this line to the beginning of the next line

File: src/forms/SelectionList/SelectionList.jsx [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+					{options.map(({ selected, label, value: optionValue, children, onClick}, index) => {
+						const elementId = `${id}-${index}`;
+						const checked = currentValue === optionValue ? true : false;
+						return (
+							<li 
+								data-testid={`${dataTestId}-option`} 
+								key={index} 
+								data-checked={checked} 
+								className={`${CSS.item} ${checked ? CSS.checked : ''}`} 
+								onClick={event => handleOnClick({ selected, label, value: optionValue, children, onClick }, index)}
```
Comment (jasonadt) [inline src/forms/SelectionList/SelectionList.jsx:49]: Please add the missing curly braces around this arrow-function body.

File: src/forms/TextFieldRule/TextFieldRule.jsx [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+					errorLabel={errorLabel} 
+					onChange={handleOnChange} 
+					onBlur={handleOnBlur} 
+					maxLength={rule} 
+					startFocused={startFocused} 
+					{...props} 
+				/>
+				<Text 
+					dataTestId={`${dataTestId}-rule`} 
+					color={`ash`}
```
Comment (jasonadt) [inline src/forms/TextFieldRule/TextFieldRule.jsx:56]: (nit) please simplify this line because it doesn't need the template and backtick syntax:

```js
color="ash"
```

File: src/index.js [hunk 20]
```diff
@@ -1,28 +1,28 @@
...
-export { default as Scroll } from './structure/Scroll'
-export { default as Wrapper } from './structure/Wrapper'
-export { default as Heading } from './typography/Heading'
-export { default as Text } from './typography/Text'
-export { default as TextField } from './forms/TextField'
-export { default as TextFieldRule } from './forms/TextFieldRule'
-export { default as TopBar } from './navigation/TopBar'
-export { default as DateField } from './forms/DateField'
-export { default as CopyableList } from './lists/CopyableList'
+export { default as ActionItem } from './lists/ActionItem/ActionItem';
```
Comment (jasonadt) [inline src/index.js:1]: 👍 thanks for sorting these! 😃

File: src/lists/ContentItem/ContentItem.jsx [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+
+ContentItem.propTypes = {
+	mediaColor: PropTypes.boolean, 
+	media: PropTypes.node, 
+	meta: PropTypes.node, 
+	onClick: PropTypes.func, 
+	className: PropTypes.string, 
+	title: PropTypes.string, 
+	dataTestId: PropTypes.string 
+
```
Comment (jasonadt) [inline src/lists/ContentItem/ContentItem.jsx:51]: (nit, no empty lines) please remove this empty line.

File: src/overlays/Dialog/Dialog.jsx [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+								className={CSS.action}
+							>
+								{confirm}
+							</button>
+						)}
+					</div>
+				</div>
+			</section>
+		)}, [visible, title, description, dismiss, children, confirm]
+	);
```
Comment (jasonadt) [inline src/overlays/Dialog/Dialog.jsx:60]: (nit) please move this to the end of the previous line (and unindent everything from lines: 19 through 59).

File: src/structure/Screen/Screen.jsx [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+
+const Screen = (props) => {
+	const [lastClick, setLastClick] = useState({x:0, y:0})
+	const viewport = { width: 0, height: 0};
+
+	useEffect(() => {
+		viewport.width = window.innerWidth;
+		viewport.height = window.innerHeight;
+		window.addEventListener('resize', handleOnWindowResize);
+		return () => {
```
Comment (jasonadt) [inline src/structure/Screen/Screen.jsx:14]: (simple return) please refactor this so that you assign this expression to a new `const result` and then separately `return result;`

File: src/structure/Screen/Screen.jsx [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+			window.removeEventListener('resize', handleOnWindowResize);
+		};
+	}, []);
+
+	const recordClick = (event) => {
+		setLastClick({x: event.clientX, y: event.clientY});
```
Comment (jasonadt) [inline src/structure/Screen/Screen.jsx:20]: (destructuring) please refactor here to use destructuring.

File: src/structure/Screen/Screen.jsx [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+	const handleOnKeyboardFocus = (screenHeight, viewport) => {
+		if (props.onKeyboardFocus) { props.onKeyboardFocus(screenHeight, viewport, lastClick) };
+	};
+
+	const handleOnKeyboardBlur = (screenHeight, viewport) => {
+		if (props.onKeyboardFocus) { props.onKeyboardBlur(screenHeight, viewport, lastClick); }
+	};
+
+	const handleOnWindowResize = event => {
+		const {target: { innerHeight: screenSize }} = event;
```
Comment (jasonadt) [inline src/structure/Screen/Screen.jsx:32]: (nit) please remove this line in favor of destructuring inside the arrow-function's signature.

File: src/forms/SelectionList/SelectionList.jsx (hunk 17 above)
Comment (jantocv) [reply src/forms/SelectionList/SelectionList.jsx:52]: 🤦 thanks

File: src/buttons/Button/Button.jsx [hunk 26]
```diff
@@ -0,0 +1,87 @@
...
+	const pressedClass = isPressed ? CSS.pressed : '';
+	const isButton = !href;
+	const defaultTestId = isButton ? 'button' : 'button-link';
+	const testId = dataTestId || defaultTestId;
+
+	if (isButton) {
+		return (
+			<button 
+				data-testid={`${testId}-button`} 
+				type="button" id={id}
```
Comment (jasonadt) [inline src/buttons/Button/Button.jsx:34]: (nit) please move the `id` attribute onto its own line

File: src/buttons/CopyToClipboardButton/Readme.md [hunk 27]
```diff
@@ -0,0 +1,29 @@
+# CopyToClipboardButton
+
+Button that copy the given value to the clipboard and provide 3 steps feedback for user
```
Comment (jasonadt) [inline src/buttons/CopyToClipboardButton/Readme.md:3]: (sp) should be: `A button that copies...` and `... and provides 3 steps of feedback to the user.`

File: src/buttons/CopyToClipboardButton/Readme.md [hunk 28]
```diff
@@ -0,0 +1,29 @@
...
+
+- Copy (default status)
+- Copying... (work in progress)
```
Comment (jasonadt) [inline src/buttons/CopyToClipboardButton/Readme.md:6]: (nit) please change `(work in progress)` to: `(Working/In progress)` which is more clear

File: src/forms/SelectionList/SelectionList.jsx [hunk 29]
```diff
@@ -0,0 +1,109 @@
...
+								<div className={CSS.content}>
+									{(label) ? (
+										<label 
+											data-testid={`${dataTestId}-option-label`} 
+											className={CSS.label} 
+											htmlFor={elementId}
+										>
+											{label}
+										</label>
+									) : children ? (
```
Comment (jasonadt) [inline src/forms/SelectionList/SelectionList.jsx:69]: (nit) please simplify this ternary operation by using `||` like so: 

```js
) : (children || null)}
```

File: src/buttons/CopyToClipboardButton/Readme.md (hunk 27 above)
Comment (jantocv) [reply src/buttons/CopyToClipboardButton/Readme.md:3]: 🇺🇸 🇬🇧 thanks!

**Comment by jantocv** [comment]
> 👍 ship it!
Lets go!
